### PR TITLE
[SPARK-30541][TESTS] Implement KafkaDelegationTokenSuite with testRetry

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
### What changes were proposed in this pull request?
`KafkaDelegationTokenSuite` has been ignored because showed flaky behaviour. In this PR I've changed the approach how the test executed and turning it on again. This PR contains the following:
* The test runs in separate JVM in order to avoid modified security context
* The body of the test runs in `testRetry` which reties if failed
* Additional logs to analyse possible failures
* Enhanced clean-up code

### Why are the changes needed?
`KafkaDelegationTokenSuite ` is ignored.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Executed the test in loop 1k+ times in jenkins (locally much harder to reproduce).

### CODE REVIEW

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDelegationTokenSuite.scala [hunk 1]
```diff
@@ -44,75 +46,79 @@ class KafkaDelegationTokenSuite extends StreamTest with SharedSparkSession with
...
-    val topic = "topic-" + UUID.randomUUID().toString
-    testUtils.createTopic(topic, partitions = 5)
 
-    withTempDir { checkpointDir =>
-      val input = MemoryStream[String]
-
-      val df = input.toDF()
-      val writer = df.writeStream
-        .outputMode(OutputMode.Append)
+      if (KafkaDelegationTokenSuite.cnt == 0) {
```
Comment (gaborgsomogyi) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDelegationTokenSuite.scala]: This is just to simulate retry. Later it will be deleted.

**Comment by gaborgsomogyi** [comment]
After local testing I've seen an interesting phenomenon. The first round of test executed properly and then the hardcoded `IllegalArgumentException` fired. Then `KDC` and Zookeeper server started properly but when `KafkaZkClient` must authenticate Kerberos is not doing a brand new authentication but found the `TGT` from the previous run (which is generated by the previous `KDC` and won't be accepted by the new):
```
Found ticket for zookeeper/localhost@EXAMPLE.COM to go to krbtgt/EXAMPLE.COM@EXAMPLE.COM expiring on Thu Mar 12 02:53:17 PDT 2020
Found ticket for zkclient/localhost@EXAMPLE.COM to go to krbtgt/EXAMPLE.COM@EXAMPLE.COM expiring on Thu Mar 12 02:53:02 PDT 2020
```
See the 15 seconds timegap between the 2, `zookeeper` `TGT` just created but `zkclient` found in the JVM's shared state. Honestly not sure why only `zkclient` `TGT` found?!

At the moment no idea how to reset this shared state and analysing the JVM source code. If somebody has idea then feel free to shed some lights...

**Comment by gaborgsomogyi** [comment]
cc @zsxwing since you've also started to reproduce this issue. Maybe you find my part useful.

**Comment by zsxwing** [comment]
@gaborgsomogyi Great investigation. Maybe we can look at Kafka's own token tests to see if they have solved the same issue.

**Comment by gaborgsomogyi** [comment]
> @gaborgsomogyi Great investigation. Maybe we can look at Kafka's own token tests to see if they have solved the same issue.

Great idea, until now I've understood `Krb5LoginModule` control flags as a whole but not yet found the cure.

**Comment by zsxwing** [comment]
One workaround right now is adding this test to this group to use its own jvm: https://github.com/apache/spark/blob/master/project/SparkBuild.scala#L463

**Comment by gaborgsomogyi** [comment]
I've just double checked `ZookeeperServer` and seems like implemented it's own Kerberos authentication manually. Tried to mimic the same steps and clear the shared state in the following way without success:
```
      System.setProperty(JAVA_AUTH_CONFIG, jaasConfigFile)
      Configuration.getConfiguration.refresh()
      val lcs = new LoginContext("Server")
      lcs.login()
      lcs.getSubject().getPublicCredentials().clear();
      lcs.getSubject().getPrivateCredentials().clear();
      val lcc = new LoginContext("Client")
      lcc.login()
      lcc.getSubject().getPublicCredentials().clear();
      lcc.getSubject().getPrivateCredentials().clear();
      val lck = new LoginContext("KafkaServer")
      lck.login()
      lck.getSubject().getPublicCredentials().clear();
      lck.getSubject().getPrivateCredentials().clear();
```

Now checking to execute it in a separate JVM but not sure how it will behave with `testRetry`. If I undertand the code properly retries will be executed in the exact same separate JVM which is not good for us, right?

**Comment by zsxwing** [comment]
Do we still need `testRetry` when running in a separate JVM?

**Comment by gaborgsomogyi** [comment]
Kafka broker may not come up since it's flaky so I think we need some kind of retry to protect us. Actually this is one of the main cause of flakyness...

**Comment by gaborgsomogyi** [comment]
`KafkaTestUtils.setup` creates an embedded broker (+KDC +zookeeper) and the broker sometimes simply not coming up (blows up in `waitUntilMetadataIsPropagated`).

**Comment by gaborgsomogyi** [comment]
Double checked the Kafka tests with the guys and haven't found test which is secure and retries within one JVM. Maybe external JVM with retry could be a solution.

**Comment by gaborgsomogyi** [comment]
@zsxwing I've changed to a new approach, namely:
* execute the suite in separate JVM => this will reduce (but not fully solve) `beforeAll` related issues. Newly created JVM mustn't have any old authentication information in place.
* use `testRetry("Roundtrip", 3)` => this will solve test related flakyness

Let's see how it behaves and I'll kick this couple of hundred times...

The bulletproof solution would be to use separate JVM with full suite retry but haven't found solution to this.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDelegationTokenSuite.scala [hunk 2]
```diff
@@ -95,6 +95,11 @@ class KafkaDelegationTokenSuite extends StreamTest with SharedSparkSession with
       }
     }
 
+    if (KafkaDelegationTokenSuite.cnt == 0) {
```
Comment (gaborgsomogyi) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDelegationTokenSuite.scala]: This is again to trigger retry and will be removed.

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
looks unrelated

**Comment by gaborgsomogyi** [comment]
retest this plese

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
seems unrelated

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
The latest code has been executed 1k+ times without failure. What do you think guys? Should we clean this approach up and merge it or do you have better approach?

@zsxwing @HyukjinKwon @dongjoon-hyun @HeartSaVioR

**Comment by HeartSaVioR** [comment]
That's a great news! I'm OK to go forward.

**Comment by HyukjinKwon** [comment]
Yes, it's okay from my side. Looks good.

**Comment by gaborgsomogyi** [comment]
Thanks guys, proceeding with it...

**Review by HeartSaVioR — APPROVED** [review]
LGTM, assuming this approach is the same being succeed for 1000 times.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 3]
```diff
@@ -74,6 +74,7 @@ class KafkaTestUtils(
   logInfo(s"Local host name is $localCanonicalHostName")
 
   private var kdc: MiniKdc = _
+  private var kdcDataDir: File = _
```
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Hm, quick question. Is this related to the fix? I couldn't follow why we need to have this var.

File: external/kafka-0-10-sql/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -25,4 +25,6 @@ log4j.appender.file.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{
 
 # Ignore messages below warning level from Jetty, because it's a bit verbose
 log4j.logger.org.spark-project.jetty=WARN
-
+log4j.logger.org.apache.spark.sql.kafka010.KafkaTestUtils=DEBUG
+log4j.logger.org.apache.directory.server.kerberos.kdc.authentication=DEBUG
+log4j.logger.org.apache.directory.server.core.DefaultDirectoryService=DEBUG
```
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/test/resources/log4j.properties]: Another quick question. How much noisy is it? I haven't tried it by myself yet but seems a bit overkill.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala (hunk 3 above)
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: consider `createKeytabsAndJaasConfigFile` as setup and `teardown` as cleanup - you'll find the reason. There might be some utility method to do cleanup of temp dir automatically as well.

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Thought the cleanup was not fully done and now I've fixed it. I've just double checked and `Utils.createTempDir()` adds this so the clean part is not needed: `ShutdownHookManager.registerShutdownDeleteDir(dir)`. Rolling back...

File: external/kafka-0-10-sql/src/test/resources/log4j.properties (hunk 4 above)
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/resources/log4j.properties]: Without these 3 loggers the output is `2836` lines and with them `3116`. I think it's a feasible addition. Having no/missing spots in the debug logs it's pretty hard to debug failures.

Comment (dongjoon-hyun) [reply external/kafka-0-10-sql/src/test/resources/log4j.properties]: Thanks. Since this is a test resource, it looks safe to me.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you all! I agree that the separate JVM and retry is the most robust option we can have here. Let's merge this and see the behavior. If we hit the flakiness again, we may need to rethink from the ground.

Merged to master/3.0.

**Comment by HyukjinKwon** [comment]
LGTM too
